File: backend/services/comp_scoring.py
"""
Similarity scoring engine for comparable properties.

Scores each candidate against the subject property across multiple dimensions.
Higher score = more similar = less adjustment needed.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any


# ── Dimension weights ──
DEFAULT_WEIGHTS = {
    "distance":       0.20,
    "recency":        0.15,
    "size":           0.20,
    "bedrooms":       0.10,
    "age":            0.05,
    "epc":            0.10,
    "imd":            0.10,
    "tier":           0.05,
    "completeness":   0.05,
}

# Tier score lookup
TIER_SCORES = {1: 1.0, 2: 0.75, 3: 0.50, 4: 0.25}


@dataclass
class ScoreBreakdown:
    """Per-dimension scores (0-1) and the composite result."""
    distance:     float | None = None
    recency:      float | None = None
    size:         float | None = None
    bedrooms:     float | None = None
    age:          float | None = None
    epc:          float | None = None
    imd:          float | None = None
    tier:         float | None = None
    completeness: float | None = None
    composite:    float = 0.0

    def to_dict(self) -> dict[str, float | None]:
        return {
            "distance": self.distance,
            "recency": self.recency,
            "size": self.size,
            "bedrooms": self.bedrooms,
            "age": self.age,
            "epc": self.epc,
            "imd": self.imd,
            "tier": self.tier,
            "completeness": self.completeness,
            "composite": self.composite,
        }


def _safe_val(obj: Any, key: str) -> float | int | None:
    """Extract a numeric value from a dict or pydantic model."""
    if isinstance(obj, dict):
        return obj.get(key)
    return getattr(obj, key, None)


def _dim_distance(comp: Any, _subj: Any) -> float | None:
    d = _safe_val(comp, "distance_m")
    if d is None:
        return None
    return max(0.0, 1.0 - min(d / 2000.0, 1.0))


def _dim_recency(comp: Any, _subj: Any) -> float | None:
    m = _safe_val(comp, "months_ago")
    if m is None:
        return None
    return max(0.0, 1.0 - min(m / 36.0, 1.0))


def _dim_size(comp: Any, subj: Any) -> float | None:
    c = _safe_val(comp, "floor_area_sqm")
    s = _safe_val(subj, "floor_area_sqm")
    if c is None or s is None or s <= 0:
        return None
    return max(0.0, 1.0 - min(abs(c - s) / s, 1.0))


def _dim_bedrooms(comp: Any, subj: Any) -> float | None:
    c = _safe_val(comp, "bedrooms")
    s = _safe_val(subj, "bedrooms")
    if c is None or s is None:
        return None
    return max(0.0, 1.0 - min(abs(c - s) / 3.0, 1.0))


def _dim_age(comp: Any, subj: Any) -> float | None:
    c = _safe_val(comp, "build_year") or _safe_val(comp, "construction_age_best")
    s = _safe_val(subj, "build_year") or _safe_val(subj, "construction_age_best")
    if c is None or s is None:
        return None
    return max(0.0, 1.0 - min(abs(c - s) / 100.0, 1.0))


def _dim_epc(comp: Any, subj: Any) -> float | None:
    c = _safe_val(comp, "epc_score")
    s = _safe_val(subj, "epc_score")
    if c is None or s is None:
        return None
    return max(0.0, 1.0 - min(abs(c - s) / 50.0, 1.0))


def _dim_imd(comp: Any, subj: Any) -> float | None:
    c = _safe_val(comp, "imd_decile")
    s = _safe_val(subj, "imd_decile")
    if c is None or s is None:
        return None
    return max(0.0, 1.0 - min(abs(c - s) / 5.0, 1.0))


def _dim_tier(comp: Any, _subj: Any) -> float | None:
    t = _safe_val(comp, "geographic_tier")
    if t is None:
        return None
    return TIER_SCORES.get(int(t), 0.25)


def _dim_completeness(comp: Any, _subj: Any) -> float:
    fields = ["floor_area_sqm", "epc_score", "imd_decile", "bedrooms", "build_year"]
    present = sum(1 for f in fields if _safe_val(comp, f) is not None)
    return present / len(fields)


# Map dimension names to scorer functions
_SCORERS = {
    "distance":     _dim_distance,
    "recency":      _dim_recency,
    "size":         _dim_size,
    "bedrooms":     _dim_bedrooms,
    "age":          _dim_age,
    "epc":          _dim_epc,
    "imd":          _dim_imd,
    "tier":         _dim_tier,
    "completeness": _dim_completeness,
}


def score_comparable(
    comp: Any,
    subject: Any,
    weights: dict[str, float] | None = None,
) -> ScoreBreakdown:
    """Score a single comparable against the subject property.

    Args:
        comp: Comparable candidate (dict or pydantic model).
        subject: Subject property (dict or pydantic model).
        weights: Optional weight overrides per dimension.

    Returns:
        ScoreBreakdown with per-dimension and composite scores.
    """
    w = weights or DEFAULT_WEIGHTS
    breakdown = ScoreBreakdown()

    total_weight = 0.0
    weighted_sum = 0.0

    for dim_name, scorer in _SCORERS.items():
        raw_score = scorer(comp, subject)
        setattr(breakdown, dim_name, raw_score)

        dim_weight = w.get(dim_name, 0.0)
        if raw_score is None:
            # Missing data: redistribute weight
            continue
        total_weight += dim_weight
        weighted_sum += raw_score * dim_weight

    breakdown.composite = weighted_sum / total_weight if total_weight > 0 else 0.0
    return breakdown


def score_pool(
    comparables: list[Any],
    subject: Any,
    weights: dict[str, float] | None = None,
) -> list[tuple[Any, ScoreBreakdown]]:
    """Score all comparables in a pool and return sorted by composite score (desc).

    Returns:
        List of (comparable, ScoreBreakdown) tuples, highest score first.
    """
    scored = [
        (comp, score_comparable(comp, subject, weights))
        for comp in comparables
    ]
    scored.sort(key=lambda x: x[1].composite, reverse=True)
    return scored

File: frontend/src/lib/comp-scoring.ts
/**
 * Comparable similarity scoring engine (frontend mirror).
 *
 * Mirrors backend/services/comp_scoring.py for real-time UI feedback
 * before the full MC simulation runs server-side.
 */

export interface ScoreBreakdown {
  distance: number | null;
  recency: number | null;
  size: number | null;
  bedrooms: number | null;
  age: number | null;
  epc: number | null;
  imd: number | null;
  tier: number | null;
  completeness: number | null;
  composite: number;
}

export interface ScoredComparable<T = any> {
  comparable: T;
  score: ScoreBreakdown;
}

const DEFAULT_WEIGHTS: Record<string, number> = {
  distance: 0.2,
  recency: 0.15,
  size: 0.2,
  bedrooms: 0.1,
  age: 0.05,
  epc: 0.1,
  imd: 0.1,
  tier: 0.05,
  completeness: 0.05,
};

const TIER_SCORES: Record<number, number> = { 1: 1.0, 2: 0.75, 3: 0.5, 4: 0.25 };

function val(obj: any, key: string): number | null | undefined {
  if (!obj) return null;
  return obj[key] ?? null;
}

function dimDistance(comp: any): number | null {
  const d = val(comp, "distance_m");
  if (d == null) return null;
  return Math.max(0, 1 - Math.min(d / 2000, 1));
}

function dimRecency(comp: any): number | null {
  const m = val(comp, "months_ago");
  if (m == null) return null;
  return Math.max(0, 1 - Math.min(m / 36, 1));
}

function dimSize(comp: any, subj: any): number | null {
  const c = val(comp, "floor_area_sqm");
  const s = val(subj, "floor_area_sqm");
  if (c == null || s == null || s <= 0) return null;
  return Math.max(0, 1 - Math.min(Math.abs(c - s) / s, 1));
}

function dimBedrooms(comp: any, subj: any): number | null {
  const c = val(comp, "bedrooms");
  const s = val(subj, "bedrooms");
  if (c == null || s == null) return null;
  return Math.max(0, 1 - Math.min(Math.abs(c - s) / 3, 1));
}

function dimAge(comp: any, subj: any): number | null {
  const c = val(comp, "build_year") ?? val(comp, "construction_age_best");
  const s = val(subj, "build_year") ?? val(subj, "construction_age_best");
  if (c == null || s == null) return null;
  return Math.max(0, 1 - Math.min(Math.abs(c - s) / 100, 1));
}

function dimEpc(comp: any, subj: any): number | null {
  const c = val(comp, "epc_score");
  const s = val(subj, "epc_score");
  if (c == null || s == null) return null;
  return Math.max(0, 1 - Math.min(Math.abs(c - s) / 50, 1));
}

function dimImd(comp: any, subj: any): number | null {
  const c = val(comp, "imd_decile");
  const s = val(subj, "imd_decile");
  if (c == null || s == null) return null;
  return Math.max(0, 1 - Math.min(Math.abs(c - s) / 5, 1));
}

function dimTier(comp: any): number | null {
  const t = val(comp, "geographic_tier");
  if (t == null) return null;
  return TIER_SCORES[t] ?? 0.25;
}

function dimCompleteness(comp: any): number {
  const fields = ["floor_area_sqm", "epc_score", "imd_decile", "bedrooms", "build_year"];
  let present = 0;
  for (const f of fields) {
    if (val(comp, f) != null) present++;
  }
  return present / fields.length;
}

type DimScorer = (comp: any, subj: any) => number | null;

const SCORERS: Record<string, DimScorer> = {
  distance: (c) => dimDistance(c),
  recency: (c) => dimRecency(c),
  size: dimSize,
  bedrooms: dimBedrooms,
  age: dimAge,
  epc: dimEpc,
  imd: dimImd,
  tier: (c) => dimTier(c),
  completeness: (c) => dimCompleteness(c),
};

export function scoreComparable(
  comp: any,
  subject: any,
  weights: Record<string, number> = DEFAULT_WEIGHTS,
): ScoreBreakdown {
  const breakdown: ScoreBreakdown = {
    distance: null,
    recency: null,
    size: null,
    bedrooms: null,
    age: null,
    epc: null,
    imd: null,
    tier: null,
    completeness: null,
    composite: 0,
  };

  let totalWeight = 0;
  let weightedSum = 0;

  for (const [dim, scorer] of Object.entries(SCORERS)) {
    const raw = scorer(comp, subject);
    (breakdown as any)[dim] = raw;

    const w = weights[dim] ?? 0;
    if (raw == null) continue;
    totalWeight += w;
    weightedSum += raw * w;
  }

  breakdown.composite = totalWeight > 0 ? weightedSum / totalWeight : 0;
  return breakdown;
}

export function scorePool<T = any>(
  comparables: T[],
  subject: any,
  weights?: Record<string, number>,
): ScoredComparable<T>[] {
  const scored = comparables.map((comp) => ({
    comparable: comp,
    score: scoreComparable(comp, subject, weights),
  }));
  scored.sort((a, b) => b.score.composite - a.score.composite);
  return scored;
}
